helix-view/src/editor.rs

```diff
@@ -67,6 +67,96 @@ where
     )
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
+#[serde(rename_all = "kebab-case", default, deny_unknown_fields)]
+pub struct GutterConfig {
+    /// Gutter Layout
+    pub layout: Vec<GutterType>,
+    /// Options specific to the "line-numbers" gutter
+    pub line_numbers: GutterLineNumbersConfig,
+}
+
+impl Default for GutterConfig {
+    fn default() -> Self {
+        Self {
+            layout: vec![
+                GutterType::Diagnostics,
+                GutterType::Spacer,
+                GutterType::LineNumbers,
+            ],
+            line_numbers: GutterLineNumbersConfig::default(),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for GutterConfig {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // Accept either a simple array of strings as gutter defaults or
+        // a [editor.gutters] block
+        #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+        #[serde(untagged)]
+        enum GutterConfigFormat {
+            Simple(Vec<GutterType>),
+            Detailed(GutterConfigStub),
+        }
```

Comment:
Custom deserializer added for `GutterConfig` so that it can accept the current `editor.gutters` vector of gutter names, while opening up more structured gutter config settings.

Currently, this can handle either

```toml
[editor]
gutters = ["diagnostics", "line-numbers"]
```

as well as

```toml
[editor.gutters]
layout = ["diagnostics", "line-numbers"]

[editor.gutters.line-numbers]
width = 3
```

Reply:
This has been cleaned up following suggestions to use the `#[serde(deserialize_with)]` attribute and a Visitor implementation. I no longer have concerns about the implementation.